README.md
# Raincouver Precipitation Prediction
☀️🌤️⛅️🌥️☁️🌦️🌧️⛈️🌩️🌨️❄️

Author: Dan Zhang, Doris (Yun Yi) Cai, Hayley (Yi) Han & Sivakorn (Oak) Chong

## Project Overview 

Our project investigates the prediction of daily precipitation in Vancouver using machine learning methods. Using a dataset spanning from 1990 to 2023, we explored the predictive power of some key environmental and climate features such as temperature, wind speed, and evapotranspiration. Our results suggest the best classification model is the Support Vector Machine with Radial Basis Function (SVM RBF) model with the hyperparameter C=10.0. The model achieved a notable F1 score of 0.87 on the positive class (precipitation is present) when generalized to the unseen data, suggesting high accuracy in precipitation prediction. We also explored feature importance, showing ET₀ reference evapotranspiration and the cosine transformation of months as robust predictors. Hyperparameter optimization did not improve our current model, indicating the potential need for feature engineering or incorporating more features. Our project presents a reliable model for predicting precipitation with potential practical applications in various fields.

The dataset we used in this project contains daily precipitation information in Vancouver from 1990 to the present (i.e., 6 Nov 2023). It is sourced from Open-Meteo’s Historical [Weather API](https://doi.org/10.5281/ZENODO.7970649) [1]. Each row in the dataset includes weather measurement statistics in a day. The key measurements in our dataset are month, daily temperature measures, wind speeds, wind direction, shortwave radiation, and ET₀ reference evapotranspiration. Specifically, shortwave radiation represents the sum of solar energy received in a day; ET₀ reference evapotranspiration provides an indication of the atmospheric demand for moisture (i.e., higher relative humidity reduces ET₀); and month is also included as a variable since it accounts for the seasonal variations in precipitation [2].

## Report

The final report is available [page](https://ubc-mds.github.io/RaincouverPrediction/).

## Dependencies

- Docker is used in this project for software dependencies management. The Docker image for this this project is built up on `quay.io/jupyter/minimal-notebook:2023-11-19`. In the [Dockerfile](https://github.com/UBC-MDS/RaincouverPrediction/blob/main/Dockerfile), it specifies additional packages and dependencies required for this project.

## Usage

#### Setup:

1. Clone the GitHub repository for this project.
   
2. [Install](https://www.docker.com/get-started/) and launch Docker on local computer and keep it run.

#### Analysis:

3. On terminal, navigate to the project local root directry and run the following command to start and run the container. Note: please ensure your localhost port 8888:8888 is not occupied before you run the command above.
   ```
   docker compose up
   ```
   
4. To bring up the web app for Jupyter Notebook in the container, look for the url starting with `http://127.0.0.1:8888/lab?token=` in terminal and copy it to browser.

5. To run the analysis, using the following commands in the terminal in the project root:

```
# download and extract data
python scripts/download_data.py \
        --url="https://archive-api.open-meteo.com/v1/archive" \
        --start-date=1990-01-01 \
        --end-date=2023-11-06\
        --write-to="./data"

# perform eda and save plots
python scripts/eda.py  \
  --data-file=data/van_weather_1990-01-01_2023-11-06.csv \
  --plot-to=results/figures

# data preprocess，split data into train and test sets,
# and save preprocessor
python scripts/drop_split_preprocess.py \
  --data-file=data/van_weather_1990-01-01_2023-11-06.csv \
  --data-to=data/processed  \
  --preprocessor-to=results/models \
  --seed=522

# model selection, model evaluation on test data and save results
python scripts/classification.py \
    --x_train=data/processed/X_train.csv \
    --y_train=data/processed/y_train.csv \
    --x-test=data/processed/X_test.csv \
    --y-test=data/processed/y_test.csv \
    --preprocessor=results/models/precipit_preprocessor.pickle \
    --columns-to-drop=parameter/columns_to_drop.csv \
    --pipeline-to=results/models \
    --plot-to=results/figures \
    --seed=522

# build HTML report and copy build to docs folder
jupyter-book build reports/milestone4
cp -r reports/milestone3/_build/html/* docs
```

6. (Alternative) To run the analysis, using the following commands(one or more) in the terminal in the project root:
```
# Run the whole analysis
make all

# Only download the data
make download_data

# Only the EDA
make eda

# Only preprocess
make drop_split_preprocess

# Only classification
make classification

# Remove the analysis
make clean
```


#### Exit container:

7. In terminal, hit `Cntrl` + `C` to stop running the container. Then use the following command to remove the container:
   ```
   docker compose rm -f
   ```

## Developer notes

#### Adding a new dependency
1. Create a new branch and add new dependency to the [Dockerfile](https://github.com/UBC-MDS/RaincouverPrediction/blob/main/Dockerfile) file, re-build the Docker image using the following commmand to ensure it works properly.
```
docker compose up
```

2. Once the updated `Dockerfile` is pushed to GitHub, a new docker image and tag created by the new `Dockerfile` will be synchronized and publish to Docker Hub automatically as the workflow has been set up on Github repository for this project.
 
3. Update the image acorrdingly in the [docker-compose.yml](https://github.com/UBC-MDS/RaincouverPrediction/blob/main/docker-compose.yml) file and ensure the container launches properly

4. Commit the changes on `docker-compose.yml` to the project repo.

5. Send PR to merge changes to 'main' branch.

#### Running the tests
Navigate to the project root directory and use the following command in terminal to test the [functions](https://github.com/UBC-MDS/RaincouverPrediction/tree/main/src) defined in the projects. Tests are stored in [here](https://github.com/UBC-MDS/RaincouverPrediction/tree/main/test).
```
pytest test/*
```

## License

The Raincouver Precipitation Prediction report contained herein are licensed under the [Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0) License](https://creativecommons.org/licenses/by-nc-sa/4.0/). The software code are licensed under [MIT License](https://opensource.org/license/mit/). If re-using/re-mixing please provide attribution and link to this webpage. See [the license file](LICENSE) for more information.


## Reference

[1] Zippenfenig, P. (2023). Open-Meteo.com Weather API [Computer software]. Zenodo. [https://doi.org/10.5281/ZENODO.7970649](https://doi.org/10.5281/ZENODO.7970649)

[2] Pal, Jeremy S., Eric E. Small, and Elfatih AB Eltahir. "Simulation of regional‐scale water and energy budgets: Representation of subgrid cloud and precipitation processes within RegCM." Journal of Geophysical Research: Atmospheres 105.D24 (2000): 29579-29594.


CHANGELOG.md
1. **Cleaned up branches**: We noticed many archived branches that were no longer in use. This is resolved in issue [#97](https://github.com/UBC-MDS/RaincouverPrediction/issues/97)
2. **Verified tests**: We lost marks in Milestone 2 because the pytest was not executed properly. This was resolved in Milestone 3 and verified in Milestone 4, as detailed in issue [#94](https://github.com/UBC-MDS/RaincouverPrediction/issues/94).
3. **Added docstrings to scripts**: Following peer review feedback about missing docstrings, we addressed this in issue [#92](https://github.com/UBC-MDS/RaincouverPrediction/issues/92)
4. **Changed folder structure**: Based on peer review feedback indicating that our folder structure was a bit confusing, we made changes in issue [#88](https://github.com/UBC-MDS/RaincouverPrediction/issues/88)
5. **Added Creative Commons License for Report**: Responding to peer review comments about the missing Creative Commons License for our report, we added one in [#90](https://github.com/UBC-MDS/RaincouverPrediction/issues/90)
6. **Adjusted Figures**: To address peer review feedback about figures being cut off in the reports, we made corrections in [#86](https://github.com/UBC-MDS/RaincouverPrediction/issues/86)
7. **Edited the references**: We updated our references to include missing DOIs as indicated in the peer review. This is resolved in [#89](https://github.com/UBC-MDS/RaincouverPrediction/issues/89)
8. **Updated Contributing.md**: We realized that our GitHub usernames were not listed in the founding members section. This has been added in [#93](https://github.com/UBC-MDS/RaincouverPrediction/issues/93)
9. **Fixed typos in the report**: Following peer review feedback about typos in our report, we addressed these in [#85](https://github.com/UBC-MDS/RaincouverPrediction/issues/85)
10. **Added more analysis in the report**: In response to peer review feedback about the need to explain our use of the f1 score, we added reasoning and elaborated on why we have the table, using recall and precision to provide further explanation. This update is in issue [#87](https://github.com/UBC-MDS/RaincouverPrediction/issues/87).
11. **Updated README**: We revised the README to reflect all the above changes. This update is detailed in issue [#91](https://github.com/UBC-MDS/RaincouverPrediction/issues/91)


scripts/classification.py
import click
import os
import altair as alt
import numpy as np
import pandas as pd
import pickle
from sklearn import set_config
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import fbeta_score, make_scorer
from joblib import dump
import sys
import matplotlib.pyplot as plt

from sklearn.metrics import classification_report
from sklearn import set_config

sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
from src.utils import cross_val_model

@click.command()
@click.option('--x_train', type=str, help="Path to X training data")
@click.option('--y_train', type=str, help="Path to y training data")
@click.option('--x-test', type=str, help="Path to x test data")
@click.option('--y-test', type=str, help="Path to y test data")
@click.option('--preprocessor', type=str, help="Path to preprocessor object", default=None)
@click.option('--columns-to-drop', type=str, help="Optional: columns to drop")
@click.option('--pipeline-to', type=str, help="Path to directory where the pipeline object will be written to")
@click.option('--plot-to', type=str, help="Path to directory where the plot will be written to")
@click.option('--seed', type=int, help="Random seed", default=123)

def main(x_train, y_train, x_test, y_test, preprocessor, columns_to_drop,
         pipeline_to, plot_to, seed): 
    """
    Trains classifiers on the provided dataset, evaluates their performance, 
    and saves the best performing model and relevant plots.

    This function reads training and test datasets, optionally applies preprocessing,
    compares multiple classifiers, and performs hyperparameter optimization on the best model. 
    It saves the optimized pipeline and generates plots showing feature importance and model comparison.

    Parameters
    ----------
    x_train : str
        Path to the CSV file containing X training data.
    y_train : str
        Path to the CSV file containing y training data.
    x_test : str
        Path to the CSV file containing X test data.
    y_test : str
        Path to the CSV file containing y test data.
    preprocessor : str
        Path to the preprocessor object. Default is None.
    columns_to_drop : str
        Optional: Path to CSV file containing names of columns to drop.
    pipeline_to : str
        Path to the directory where the pipeline object will be saved.
    plot_to : str
        Path to the directory where plots will be saved.
    seed : int
        Random seed for reproducibility. Default is 123.

    Returns
    -------
    None
        This function does not return any value but writes the optimized pipeline object 
        and plots to the specified paths.

    Notes
    -----
    - The function uses a variety of classifiers including Decision Tree, KNN, SVM, and Logistic Regression.
    - Hyperparameter optimization is performed on the best model based on the F1 score.
    - The function generates plots for feature importance, model comparison, and a classification report.
    - The function will exit if the best model is not SVM RBF, as it is set up to optimize this specific model.

    Examples
    --------
    Command line usage:
    $ python scripts/classification.py \
             --x_train=data/processed/X_train.csv \
             --y_train=data/processed/y_train.csv \
             --x-test=data/processed/X_test.csv \
             --y-test=data/processed/y_test.csv \
             --preprocessor=results/models/precipit_preprocessor.pickle \
             --columns-to-drop=data/processed/columns_to_drop.csv \
             --pipeline-to=results/models \
             --plot-to=results/figures \
             --seed=522
    """

    np.random.seed(seed)
    set_config(transform_output="pandas")

    #Read in data and preprocessor
    X_train = pd.read_csv(x_train)
    y_train = pd.read_csv(y_train)
    
    y_col_name = y_train.columns.tolist()[0]
    y_train_class = y_train[y_col_name]
    preprocess = pickle.load(open(preprocessor, "rb"))
    
    if columns_to_drop:
        to_drop = pd.read_csv(columns_to_drop).feats_to_drop.tolist()
        X_train = X_train.drop(columns=to_drop)

    click.echo(f'Running model training for columns {X_train.columns.tolist()}')

    #Model parameter to be explored
    models = {
    "Decision Tree": DecisionTreeClassifier(random_state=522),
    "KNN": KNeighborsClassifier(),
    "RBF SVM": SVC(random_state=522),
    "Logistic Regression": LogisticRegression(max_iter=2000, multi_class="ovr", random_state=522),
    }

    #Plot out feature importance
    pipe = make_pipeline(preprocess, models["Logistic Regression"])
    pipe.fit(X_train, y_train_class)
    numeric_features = pipe.named_steps['columntransformer'].named_transformers_['standardscaler'].get_feature_names_out().tolist()
    coefficients = pipe.named_steps['logisticregression'].coef_[0]
    
    feature_importance = pd.DataFrame({'Feature': numeric_features, 'Importance': np.abs(coefficients)})
    feature_importance = feature_importance.sort_values('Importance', ascending=True)   
    plt.figure(figsize=(12, 10))
    feature_importance.plot(x='Feature', y='Importance', kind='barh')
    plt.title("Feature_importance")
    plt.tight_layout()
    plt.savefig(os.path.join(plot_to, "Feature_importance.png"))

    click.echo(f'Feature importance has been parked at {os.path.join(plot_to, "Feature_importance.png")}')

    #Model type selection
    classification_metrics = ["accuracy", "precision", "recall", "f1"]
    results_df = cross_val_model(preprocess, models, X_train, y_train_class, classification_metrics)

    #Choose the model based on the best F1 score. 
    result_dict = results_df.loc['test_f1', :].to_dict()

    # Create a bar chart with F1 scores all the tested models and publish it
    plt.figure(figsize=(10, 8))
    plt.bar(models.keys(), result_dict.values(), color='blue')
    plt.xlabel('Models', fontsize=15)
    plt.ylabel('Test F1 Score', fontsize=15)
    plt.title('Test F1 Scores for Different Models', fontsize=18)
    plt.ylim(0.8, 0.9)  # Set the y-axis limit between 0 and 1 for F1 scores
    plt.xticks(rotation=45, fontsize=14)  # Rotate x-axis labels for better readability
    plt.yticks(fontsize=14) 
    plt.tight_layout()
    plt.savefig(os.path.join(plot_to, "model_comparison.png"), dpi=300, bbox_inches='tight')
    
    click.echo(f'Model F1 performance has been parked at {os.path.join(plot_to, "model_comparison.png")}')
    # Save the bar chart as an image
    plt.tight_layout()
    
    #Get model name with max_score
    max_score = max(result_dict, key=result_dict.get)
    model_name = max_score[0]

    #Echo out if the model chosen is not expected as SVM RBF
    if model_name == 'RBF SVM':
        click.echo('Proceed with hyperparameter optimization for RBF SVM model')
    else:
        click.echo('Best model is not expected SVM RBF. Other model parameters required for optimization, exitting the program')
        sys.exit(0)

    #Hyperparameter optimization
    param_grid = {"svc__C": 10.0**np.arange(-3,3)}
    svc_pipe = make_pipeline(preprocess, models[model_name])
    grid_search = GridSearchCV(svc_pipe,param_grid=param_grid,n_jobs=-1,return_train_score=True)
    grid_search.fit(X_train, y_train_class)
    C_best_value = grid_search.best_params_['svc__C']
    opt_pipe = make_pipeline(preprocess, SVC(C=C_best_value, random_state = 522))
    
    opt_pipe.fit(X_train, y_train_class)

    #Saving optimum pipe as a pickle file for testing
    with open(os.path.join(pipeline_to, "optimum_cls_svm_pipeline.pickle"), 'wb') as f:
        pickle.dump(opt_pipe, f)

    click.echo(f'Optimized model has been parked at {os.path.join(pipeline_to, "optimum_cls_svm_pipeline.pickle")}')

    #Saving model evaluation result
    X_test = pd.read_csv(x_test)
    y_test = pd.read_csv(y_test)
    y_col_name = y_test.columns.tolist()[0]
    y_test_class = y_test[y_col_name]

    if columns_to_drop:
        to_drop = pd.read_csv(columns_to_drop).feats_to_drop.tolist()
        X_test = X_test.drop(columns=to_drop)
        
    click.echo(f'Running model test for columns {X_test.columns.tolist()}')        
    y_pred = opt_pipe.predict(X_test)

    report_dict = classification_report(y_test_class, y_pred, output_dict=True)
    report_df = pd.DataFrame(report_dict).transpose()
    
    report_df = report_df.loc[['False', 'True']].round(2)
    report_df.index = ['No rain', 'Rain']
    
    plt.figure(figsize=(5, 1))  # Adjusting the figure size
    ax = plt.gca()
    table = ax.table(cellText=report_df.values,
                     colLabels=report_df.columns,
                     rowLabels=report_df.index,
                     loc='center',
                     cellLoc='center')
    table.scale(1, 1.5) 
    ax.axis('off')
    plt.tight_layout()
    plt.savefig(os.path.join(plot_to, "classification_report.png"), dpi=300, bbox_inches='tight')
    click.echo(f'Classification report has been parked at {os.path.join(plot_to, "classification_report.png")}')

if __name__ == '__main__':
    main()
